nixos/modules/services/monitoring/karma.nix

```diff
@@ -0,0 +1,179 @@
+{ config, pkgs, lib, ... }:
+with lib;
+let
+  cfg = config.services.karma;
+
+in
+{
+  options.services.karma = {
+    enable = mkEnableOption (mdDoc "the Karma dashboard service.");
+
+    package = mkOption {
+      type = types.package;
+      default = pkgs.karma;
+      defaultText = literalExpression "pkgs.karma";
+      description = mdDoc ''
+        The Karma package that should be used.
+      '';
+    };
+
+    configFile = mkOption {
+      type = with types; nullOr path;
+      default = null;
+      description = mdDoc ''
+        Provides Karma with a raw config file instead of a nix-generated file.
+      '';
+      example = "/etc/karma/karma.conf";
+    };
+
+    environment = mkOption {
+      type = types.attrs;
+      default = {};
+      description = mdDoc ''
+        Additional environment variables to provide to karma.
+      '';
+      example = {
+        ALERTMANAGER_URI = "https://alertmanager.example.com";
+        ALERTMANAGER_NAME= "single";
+      };
+    };
+
+    openFirewall = mkOption {
+      type = types.bool;
+      default = false;
+      description = mdDoc ''
+        Whether to open ports in the firewall needed for karma to function.
+      '';
+    };
+
+    extraOptions = mkOption {
+      type = with types; listOf str;
+      default = [];
+      description = mdDoc ''
+        Extra command line options.
+      '';
+      example = [
+        "--alertmanager.timeout 10s"
+      ];
+    };
+
+    address = mkOption {
+      type = types.str;
+      default = "127.0.0.1";
+      description = mdDoc ''
+        Hostname or IP to listen on.
+      '';
+      example = "[::]";
+    };
+
+    port = mkOption {
+      type = types.port;
+      default = 8080;
+      description = mdDoc ''
+        HTTP port to listen on.
+      '';
+      example = 8182;
+    };
+
+    readTimeout = mkOption {
+      type = with types; nullOr str;
+      default = null;
+      description = mdDoc ''
+        HTTP server request read timeout.
+      '';
+      example = "15s";
+    };
+
+    writeTimeout = mkOption {
+      type = with types; nullOr str;
+      default = null;
+      description = mdDoc ''
+        HTTP server response write timeout.
+      '';
+      example = "15s";
+    };
+
+    urlPrefix = mkOption {
+        type = types.str;
+        default = "/";
+        description = mdDoc ''
+          URL root for karma, you can use this to serve from a location other than `/`.
+
+          This option is mostly useful when using karma behind a reverse proxy with other services on the same IP but different URL root.
+        '';
+        example = "/dashboard";
```

Comment:
Please use consistent indentation: 2 spaces per level of indentation.

Reply:
Whoops, that's an artifact from before I first refactored it. Fixed.